### PR TITLE
[WIP] add subscription pipeline

### PR DESCRIPTION
**Pull request by lghinet**
# Description

Adds the possibility to configure a different pipeline, one for publishing and other one for subscribing to messages from a message bus

## Issue reference

Please reference the issue this PR will close: https://github.com/dapr/dapr/issues/3361

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
I do not think pubsub middleware is something we want to support over HTTP only. Filtering and transformation of pubsub events need to be designed together and have a consistent experience in Dapr. The PR in https://github.com/dapr/dapr/pull/3406 will take care of filtering on the subscriber side and we should build on top of that experience instead of HTTP Middleware.

File: charts/dapr/crds/configuration.yaml [hunk 1]
```diff
@@ -140,6 +140,84 @@ spec:
                 required:
                 - handlers
                 type: object
+              publishPipeline:
```
Comment (artursouza) [inline charts/dapr/crds/configuration.yaml:143]: This feature does not belong to HTTP Middleware. See https://github.com/dapr/dapr/pull/3406

**Review by artursouza — CHANGES_REQUESTED** [review]

File: charts/dapr/crds/configuration.yaml (hunk 1 above)
Comment (lghinet) [reply charts/dapr/crds/configuration.yaml:143]: why do you say its related to http middleware ? 
also filtering can be a middleware in this pipeline, you don't a separate mechanism for that, you can have a middleware that does the filtering and condition == false short-circuits the flow ..

Comment (artursouza) [reply charts/dapr/crds/configuration.yaml:143]: An app might decide to have multiple pipelines. Can this be exported outside configuration? Maybe inside the Subscription CRD?

Comment (lghinet) [reply charts/dapr/crds/configuration.yaml:143]: well, we have two new pipelines, one the subscriber side, and on the publisher side..  (as you can see in the configuration CRD ) 

let's say the first one can be configured in the SubscriptionCRD, (as you suggested),  but the second one .. might not even have a subscription, it's just a publisher that generates new messages into some topic, so it doesn't need a subscription CRD,
we have to find another place to put it

Comment (lghinet) [reply charts/dapr/crds/configuration.yaml]: moved the pipeline to subscription crd

**Comment by lghinet** [comment]
@artursouza  can you take a look ? .. a moved the pipeline config in the subscription crd

**Comment by yaron2** [comment]
Thanks for this PR @lghinet, this is something @artursouza and I should look at closer in terms of the CRD design that's suggested here. In the meantime, can you resolve the conflicts?

**Comment by lghinet** [comment]
hey @yaron2, this merge is not that easy because a lot of changes have occurred in the master branch, so I think it's better that you guys reach a conclusion first, then I will redo the task,  
otherwise I risk redoing this many times, which is quite time consuming .. 
thanks

**Comment by lghinet** [comment]
> Thanks for this PR @lghinet, this is something @artursouza and I should look at closer in terms of the CRD design that's suggested here. In the meantime, can you resolve the conflicts?

done,

**Pull request closed without merging**